typegen/src/utils.rs

```diff
@@ -0,0 +1,130 @@
+use scale_info::{form::PortableForm, Field, PortableRegistry, Type, TypeDef, TypeParameter};
+use smallvec::{smallvec, SmallVec};
+use std::collections::HashMap;
+
+pub fn ensure_unique_type_paths(types: &mut PortableRegistry) {
```

Comment:
This is probably worth a comment to describe the algorithm we're using, because it's a bit tricky to follow!

I had a go to see if I could simplify it and came up with this offhand (just avoiding some intermediate allocs/steps and doing all grouping and then all renaming in 2 steps):

```rust
let mut types_with_same_type_path_grouped_by_shape = HashMap::<&[String], Vec<Vec<u32>>>::new();

// First, group types if they are similar (same path, same shape).
for (ty_id, ty) in types.types.iter().enumerate() {
    // Ignore types without a path (i.e prelude types).
    if ty.ty.path.namespace().is_empty() {
        continue;
    };

    // get groups that share this path already, if any.
    let mut groups_with_same_path = types_with_same_type_path_grouped_by_shape
        .entry(&ty.ty.path.segments)
        .or_default();
    
    // Compare existing groups to check which to add our type ID to.
    let mut added_to_existing_group = false;
    for group in &mut groups_with_same_path {
        let ty_id_b = group[0]; // all types in group are same shape; just check any one of them.
        let ty_b = types.resolve(ty_id_b).expect("ty exists");
        if types_equal_extended_to_params(&ty, ty_b) {
            group.push(ty_id_b);
            added_to_existing_group = true;
            break;
        }
    }

    // We didn't find a matching group, so add it to a new one.
    if !added_to_existing_group {
        groups_with_same_path.push(vec![ty_id])
    }
}

// Now, rename types as needed based on these groups.
let groups_that_need_renaming = types_with_same_type_path_grouped_by_shape
    .values()
    .filter(|g| g.len() > 1);

for groups_with_same_path in groups_that_need_renaming {
    let mut n = 1;
    for group_with_same_shape in groups_with_same_path {
        for ty_id in group_with_same_shape {
            let ty = types
                .types
                .get_mut(ty_id as usize)
                .expect("type is present; qed;");
            let name = ty.ty.path.segments.last_mut().expect("This is only empty for builtin types, that are filtered out with namespace().is_empty() above; qed;");
            *name = format!("{name}{appendix}"); // e.g. Header1, Header2, Header3, ...
        }
        n += 1;
    }
}
```